backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CpuPinningPolicyUnit.java

```diff
@@ -58,24 +59,33 @@ public CpuPinningPolicyUnit(PolicyUnit policyUnit,
     @Override
     public List<VDS> filter(final SchedulingContext context,
             final List<VDS> hosts,
-            final VM vm,
+            final List<VM> vmGroup,
             final PerHostMessages messages) {
-        final String cpuPinning = vm.getVmPinning();
 
         // return all hosts when no CPU pinning is requested
-        if (vm.getCpuPinningPolicy() == CpuPinningPolicy.NONE) {
+        if (vmGroup.stream().allMatch(vm -> vm.getCpuPinningPolicy() == CpuPinningPolicy.NONE)) {
             return hosts;
         }
 
         // only add hosts as candidates which have all required CPUs up and running
-        final List<VDS> candidates = new ArrayList<>();
+        final List<VDS> candidates = new ArrayList<>(hosts);
 
-        if (vm.getCpuPinningPolicy().isExclusive()) {
-            for (final VDS host : hosts) {
+        List<VM> exclusiveVms = vmGroup.stream()
+                .filter(vm -> vm.getCpuPinningPolicy().isExclusive())
+                .sorted(new VmsCpuPinningPolicyComparator())
+                .collect(Collectors.toList());
+        List<VM> sharedVms = vmGroup.stream()
+                .filter(vm -> !vm.getCpuPinningPolicy().isExclusive())
+                .collect(Collectors.toList());
```

Comment:
Maybe you could also check if the cpuPinning is empty here and avoid doing that later
```
        if (StringUtils.isEmpty(cpuPinning)) {
            // For 'Resize and Pin NUMA', on RunVm, we get here before determining the CPU pinning for the VM
            // so the dynamic CPU pinning is not set and hosts should not be filtered by the logic below.
            return hosts;
        }
```